replacer_test.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2015 Matthew Holt and The Caddy Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caddy
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+)
+
+func fakeReplacer() replacer {
+	return replacer{
+		providers: make([]ReplacementFunc, 0),
+		static:    make(map[string]string),
+	}
+}
+
+func fakeReplacerFilled() replacer {
+	return replacer{
+		providers: []ReplacementFunc{
+			// split our possible vars to two functions (to test if both functions are called)
+			func(key string) (val string, ok bool) {
+				switch key {
+				case "test1":
+					return "val1", true
+				case "asdf":
+					return "123", true
+				case "äöü":
+					return "öö_äü", true
+				case "with space":
+					return "space value", true
+				default:
+					return "NOOO", false
+				}
+			},
+			func(key string) (val string, ok bool) {
+				switch key {
+				case "1":
+					return "test-123", true
+				case "mySuper_IP":
+					return "1.2.3.4", true
+				case "testEmpty":
+					return "", true
+				default:
+					return "NOOO", false
+				}
+			},
+		},
+	}
+}
+
+func fakeReplacerFilledStatic() replacer {
+	return replacer{
+		static: map[string]string{
+			"key1": "val1",
+			"key2": "val2",
+			"key3": "val3",
+			"key4": "val4",
+		},
+	}
+}
+
+func getHostnameForTest() string {
+	name, _ := os.Hostname()
+	return name
+}
+
+func getEnvForTest(name string) (string, bool) {
+	val := os.Getenv(name)
+	return val, val != ""
+}
+
+// Tests the Set method by setting some variables and check if they are added to static
+func TestReplacerSet(t *testing.T) {
+	rep := fakeReplacer()
+
+	for _, tc := range []struct {
+		variable string
+		value    string
+	}{
+		{
+			variable: "test1",
+			value:    "val1",
+		},
+		{
+			variable: "asdf",
+			value:    "123",
+		},
+		{
+			variable: "äöü",
+			value:    "öö_äü",
+		},
+		{
+			variable: "with space",
+			value:    "space value",
+		},
+		{
+			variable: "1",
+			value:    "test-123",
+		},
+		{
+			variable: "mySuper_IP",
+			value:    "1.2.3.4",
+		},
+		{
+			variable: "testEmpty",
+			value:    "",
+		},
+	} {
+		rep.Set(tc.variable, tc.value)
+
+		// test if key is added
+		if val, ok := rep.static[tc.variable]; ok {
```

Comment:
This might be better as two loops through the test cases:

1. Call `Set()` and do nothing else
2. Then check each variable and make sure the value is correct.

This way we can guarantee that values aren't disappearing and aren't being overwritten either when they shouldn't be.